src/controllers/deploy.go

```diff
@@ -69,20 +69,24 @@ func internaldeploy(a *history.ActionInstance) ([]byte, error) {
 		state.Access = a.Access
 		state.Server = a.Server
 		state.Status = "deploying"
-		if err1 := history.SetState(a.RepoURL, state); err1 != nil {
+		tag, err1 := history.SetState(a.RepoURL, tag, state)
+		if err1 != nil {
 			log.Infof("setting state to deploying failed - %v", err1)
 			output = []byte("InternalDeployError: cannot set state to deploying - " + err1.Error())
 			return output, err1
 		}
 		output, err = exec.Command(deployScriptName, "-n", "-u", a.RepoURL, "-b", branch, "-m", a.Server, "-s", a.Subdomain, "-a", a.Access).CombinedOutput()
 		if err != nil {
 			state.Status = "stopped"
-			history.SetState(a.RepoURL, state)
 		} else {
 			state.Status = "running"
-			history.SetState(a.RepoURL, state)
 		}
 
+		tag, err1 = history.SetState(a.RepoURL, tag, state)
+		for ; err1 != nil; tag, err1 = history.SetState(a.RepoURL, tag, state) {
```

Comment:
Why do you need a for loop? Technically this SetState should never fail, and if it does, then we should return an error, instead of running it again and again. This looks like a weird way to just forcefully SetState to something, which should not happen, because something changed in the mean time.

Reply:
This should not fail, because if the the SetState suceeded in line 72, then we get into a "-ing" state and no operation is allowed on it then.

Reply:
Yes, it should not fail. But what should happen if it ever does? If we return it, the service is left in a state, that nothing would be able to work on. Is that acceptable?

Reply:
Technically, what I would like is to put it in "error" state and then allow only stop operation on it. 

Later, when we have alerting, we can post a message on discord that something is in error state, and we manually stop it.   

For the time-being, can you log.Error instead of log.Info ?